### ISSUE TITLE
Numerical instability in Google Colab - Part 4 of Makemore

### ISSUE THREAD

**Issue by sachag678** [karpathy/nn-zero-to-hero#13]
I ran into an interesting issue in makemore 4 backpro ninja where the dhpreact was not exactly matching the hpreact.grad. 

However, this was only in the collab notebook because when I put the same code into a local jupyter notebook it works fine. 

Not sure why this would be the case but just an odd curiosity.

**Comment by karpathy** [comment]
oh oh

**Comment by sachag678** [comment]
I'm guessing it has something to do with the python versions?

**Comment by JonathanSum** [comment]
Yes. I have an issue with colab. But I don't have an issue on the local vscode Jutpyer notebook with **Pytorch 3.7.13**
![image](https://user-images.githubusercontent.com/21982975/195742046-50ba3529-803e-4df8-85ca-508dfa1661d0.png)
![image](https://user-images.githubusercontent.com/21982975/195742257-01c4b833-1832-457b-a36b-a1d7de6221f0.png)

The colab version is 
**3.7.14 (default, Sep  8 2022, 00:06:44)** 
[GCC 7.5.0]
If the diff number is too small, maybe it is fine to use some way to accept it?
Colab tested notebook: https://colab.research.google.com/drive/1HmZ8bgtAfvyMaZyu3Sr1Bgxsj35jitTs?usp=sharing

**Comment by JonathanSum** [comment]
I used the sum method to do comparing the colab difference vs the local version. But I didn't set the same generator or same seed, so the comparison may not be helpful.
[colab.txt](https://github.com/karpathy/nn-zero-to-hero/files/9781858/colab.txt)
[local.txt](https://github.com/karpathy/nn-zero-to-hero/files/9781859/local.txt)

**Comment by mriganktiwari** [comment]
> Yes. I have an issue with colab. But I don't have an issue with the local VScode Jupyter notebook. The local Jupyter notebook version is **Python 3.7.13** The tested colab notebook version is **3.7.14 (default, Sep 8 2022, 00:06:44)** [GCC 7.5.0] ![image](https://user-images.githubusercontent.com/21982975/195742833-e2c3091a-b293-4c9d-b60a-a2ec7e9ccb1a.png)
> 
> ![image](https://user-images.githubusercontent.com/21982975/195742046-50ba3529-803e-4df8-85ca-508dfa1661d0.png) ![image](https://user-images.githubusercontent.com/21982975/195742547-ea30843e-9de1-4224-9720-547b49dda6d8.png)
> 
> If the diff number is too small, maybe it is fine to use some way to accept it? Colab tested notebook: https://colab.research.google.com/drive/1HmZ8bgtAfvyMaZyu3Sr1Bgxsj35jitTs?usp=sharing
> 
> Maybe the issue is Pytorch version?

I am getting exactly same ```maxdiff``` for **hpreact**, and my notebook is running on local machine.
Python 3.9.13

&

>>> torch.__version__
'1.12.1'

**Comment by evgenyfadeev** [comment]
I've got a strange observation (using the colab version)

`dlogit_maxes =  - dnorm_logits.sum(dim=1, keepdim=True)` gives me exact equality
`dlogit_maxes =  - dnorm_logits.sum(dim=1)` gives approximate equality with a maxdiff ~ 10^-9

**Comment by vdyma** [comment]
I had the same difference problem between gradients when running locally, because I used GPU to store tensors and perform computations. Once I changed to CPU, I had the difference in the later computations because of the ordering of operations. I managed to get the exact gradients running on CPU and reordering computations to be the same as in the lecture.

**Comment by conscell** [comment]
I encountered the same issue on a Linux machine with CPU. Setting the following environment variable resolved the problem:
```
ATEN_CPU_CAPABILITY=default
```

To fix the issue in the notebook, add these lines at the very beginning of the notebook, before importing PyTorch:
```
import os
os.environ['ATEN_CPU_CAPABILITY'] = 'default'
```

However, this solution does not address issues with Nvidia GPUs, which remain affected.
